### PR TITLE
[Driver][SYCL][FPGA] Adjust the output location for the project report

### PR DESCRIPTION
**Pull request by mdtoguchi**
When no output specifier is given on the command line, create the output
project report location in the $CWD as opposed to matching the source location
if it was specified with a directory.

Signed-off-by: Michael D Toguchi <michael.d.toguchi@intel.com>

### CODE REVIEW

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -278,9 +278,14 @@ void SYCL::fpga::BackendCompiler::ConstructJob(Compilation &C,
         if (Ty == types::TY_INVALID)
           continue;
         if (types::isSrcFile(Ty) || Ty == types::TY_Object) {
-          llvm::sys::path::replace_extension(AN, "prj");
-          ReportOptArg += Args.MakeArgString(AN);
-          break;
+          // Output is based on the filename only so strip off any directory
+          // information if provided with the source/object file.
+          AN = llvm::sys::path::filename(AN);
+          if (!AN.empty()) {
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp]: Under what circumstances could `filename()` return an empty string?

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 2]
```diff
@@ -278,9 +278,14 @@ void SYCL::fpga::BackendCompiler::ConstructJob(Compilation &C,
         if (Ty == types::TY_INVALID)
           continue;
         if (types::isSrcFile(Ty) || Ty == types::TY_Object) {
-          llvm::sys::path::replace_extension(AN, "prj");
-          ReportOptArg += Args.MakeArgString(AN);
-          break;
+          // Output is based on the filename only so strip off any directory
+          // information if provided with the source/object file.
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp]: IMHO, the intent could be expressed more clearly with smth like:
```suggestion
          // Project report should be saved into CWD, so strip off any
          // directory information if provided with the input file.
```

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 3]
```diff
@@ -278,9 +278,14 @@ void SYCL::fpga::BackendCompiler::ConstructJob(Compilation &C,
...
+          AN = llvm::sys::path::filename(AN);
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp]: As an out-of-scope note, `AN` is a tough name for figuring the meaning out - am I missing some common abbreviation?

Comment (meiyacha) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Another question on AN, or rather Args. Are those always filenames? Can't Args contain other options?

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 4]
```diff
@@ -171,6 +171,16 @@
 // RUN:  | FileCheck -DOUTDIR=%t_dir -check-prefix=CHK-FPGA-REPORT-OPT %s
 // CHK-FPGA-REPORT-OPT: aoc{{.*}} "-sycl" {{.*}} "-output-report-folder=[[OUTDIR]]{{/|\\\\}}file.prj"
 
+/// -fintelfpga output report file from dir/source
```
Comment (meiyacha) [inline clang/test/Driver/sycl-offload-intelfpga.cpp]: Will this test be run on all platforms?

Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload-intelfpga.cpp]: That is the expectation, but currently is not true reality.  There are currently adjustments that are needed to have this test run on a Windows machine - so the test is run but is expected to fail.  We at least exercise what would be Windows like paths by running clang_cl on the linux host.

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 3 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: `AN` was an abbreviation of Arg Name (string of the Arg).  I'll modify to be more succinct.  As for the Args question, the Args consist of all arguments passed on the command line.  So they are not all file names.  But good catch - there should be an additional check to make sure that the Arg we are currently looking at is an Input type.

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Probably none (I can't think of any).  I'm sure I added the check due to my paranoia.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 5]
```diff
@@ -271,16 +271,22 @@ void SYCL::fpga::BackendCompiler::ConstructJob(Compilation &C,
...
-      StringRef Ext(llvm::sys::path::extension(AN));
-      if (!Ext.empty()) {
-        types::ID Ty = getToolChain().LookupTypeForExtension(Ext.drop_front());
-        if (Ty == types::TY_INVALID)
-          continue;
-        if (types::isSrcFile(Ty) || Ty == types::TY_Object) {
-          llvm::sys::path::replace_extension(AN, "prj");
-          ReportOptArg += Args.MakeArgString(AN);
-          break;
+      if (Cur->getOption().getKind() == Option::InputClass) {
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp]: Is an early exit possible here?

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 6]
```diff
@@ -271,16 +271,22 @@ void SYCL::fpga::BackendCompiler::ConstructJob(Compilation &C,
...
+        SmallString<128> ArgName = Cur->getSpelling();
+        StringRef Ext(llvm::sys::path::extension(ArgName));
+        if (!Ext.empty()) {
+          types::ID Ty =
+              getToolChain().LookupTypeForExtension(Ext.drop_front());
+          if (Ty == types::TY_INVALID)
+            continue;
+          if (types::isSrcFile(Ty) || Ty == types::TY_Object) {
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp]: If the "inputness" check above is successful, are we still concerned with the particular input type?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: No, we shouldn't be - When the Input action is created, there should always be a valid type.  I'll clean up.

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 5 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: naming is based on source or object, so we can't guarantee early exit just based on first input seen. (if that is what you were getting at)

Comment (AGindinson) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Oh, sorry for the confusion: I've rather meant an early `continue` in iterating over `Args`

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Given that there is nothing else to skip over if this condition isn't met I don't see any opportunity for an early exit.  I'm open to suggestions if you see an opportunity.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 7]
```diff
@@ -271,16 +271,20 @@ void SYCL::fpga::BackendCompiler::ConstructJob(Compilation &C,
...
+            // Project report should be saved into CWD, so strip off any
+            // directory information if provided with the input file.
+            ArgName = llvm::sys::path::filename(ArgName);
+            llvm::sys::path::replace_extension(ArgName, "prj");
+            ReportOptArg += Args.MakeArgString(ArgName);
+            break;
+          }
         }
       }
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp]: This is the "early continue" suggestion from the remaining thread that I'm talking about:
```suggestion
      if (Cur->getOption().getKind() != Option::InputClass)
        continue;
      SmallString<128> ArgName = Cur->getSpelling();
      StringRef Ext(llvm::sys::path::extension(ArgName));
      if (Ext.empty())
        continue;
      types::ID Ty =
          getToolChain().LookupTypeForExtension(Ext.drop_front());
      if (types::isSrcFile(Ty) || Ty == types::TY_Object) {   // Could also use early continue, but feels less readable smh
        // Project report should be saved into CWD, so strip off any
        // directory information if provided with the input file.
        ArgName = llvm::sys::path::filename(ArgName);
        llvm::sys::path::replace_extension(ArgName, "prj");
        ReportOptArg += Args.MakeArgString(ArgName);
        break;
      }
```

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Thanks for the suggestion.  Also found that the `TY_INVALID` check is needed as there is no guarantee that the extension is going to provide a reasonable type that we expect.

Comment (AGindinson) [reply clang/lib/Driver/ToolChains/SYCL.cpp]: Fair enough! Thanks.

**Pull request merged**